logsmanagement/flb_plugin.c

```diff
@@ -0,0 +1,982 @@
+struct flb_time {
+    struct timespec tm;
+};
+
+struct flb_lib_ctx {
+    int status;
+    struct mk_event_loop *event_loop;
+    struct mk_event *event_channel;
+    struct flb_config *config;
+};
+
+struct flb_parser_types {
+    char *key;
+    int  key_len;
+    int type;
+};
+
+struct flb_parser {
+    /* configuration */
+    int type;             /* parser type */
+    char *name;           /* format name */
+    char *p_regex;        /* pattern for main regular expression */
+    int skip_empty;       /* skip empty regex matches */
+    char *time_fmt;       /* time format */
+    char *time_fmt_full;  /* original given time format */
+    char *time_key;       /* field name that contains the time */
+    int time_offset;      /* fixed UTC offset */
+    int time_keep;        /* keep time field */
+    int time_strict;      /* parse time field strictly */
+    char *time_frac_secs; /* time format have fractional seconds ? */
+    struct flb_parser_types *types; /* type casting */
+    int types_len;
+
+    /* Field decoders */
+    struct mk_list *decoders;
+
+    /* internal */
+    int time_with_year;   /* do time_fmt consider a year (%Y) ? */
+    char *time_fmt_year;
+    int time_with_tz;     /* do time_fmt consider a timezone ?  */
+    struct flb_regex *regex;
+    struct mk_list _head;
+};
+
+struct flb_lib_out_cb {
+    int (*cb) (void *record, size_t size, void *data);
+    void *data;
+};
+
+typedef struct flb_lib_ctx flb_ctx_t;
+
+static flb_ctx_t *ctx;
+static flb_ctx_t *(*flb_create)(void);
+static int (*flb_service_set)(flb_ctx_t *ctx, ...);
+static int (*flb_start)(flb_ctx_t *ctx);
+static int (*flb_time_pop_from_msgpack)(struct flb_time *time, msgpack_unpacked *upk, msgpack_object **map);
+static int (*flb_lib_free)(void *data);
+static struct flb_parser *(*flb_parser_create)(const char *name, const char *format, const char *p_regex,int skip_empty,
+                                               const char *time_fmt, const char *time_key, const char *time_offset,
+                                               int time_keep, int time_strict, struct flb_parser_types *types,
+                                               int types_len, struct mk_list *decoders, struct flb_config *config);
+static int (*flb_input)(flb_ctx_t *ctx, const char *input, void *data);
+static int (*flb_input_set)(flb_ctx_t *ctx, int ffd, ...);
+static int (*flb_output)(flb_ctx_t *ctx, const char *output, struct flb_lib_out_cb *cb);
+static int (*flb_output_set)(flb_ctx_t *ctx, int ffd, ...);
+static msgpack_unpack_return (*dl_msgpack_unpack_next)(msgpack_unpacked* result, const char* data, size_t len, size_t* off);
+static void (*dl_msgpack_zone_free)(msgpack_zone* zone);
+
+// TODO: Update "flush", "0.1" according to minimum update every?
+int flb_init(void){
+
+    /* Load Fluent-Bit functions from the shared library */
+    void *handle;
+    char *dl_error;
+
+    handle = dlopen("/tmp/netdata/usr/lib/netdata/libfluent-bit.so", RTLD_LAZY);
```

Comment:
[This](https://github.com/bpowers/musl/blob/master/src/ldso/dlopen.c) is what I said for your. Unfortunately the whole `libdl` needs to be removed, because `musl` does not support it, and we need it a lot.

Reply:
I will mark this one as resolved, since it's been tested now and it works on Alpine. If I see any issues with it, I will investigate them.